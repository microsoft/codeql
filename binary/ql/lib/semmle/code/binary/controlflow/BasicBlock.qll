private import binary
private import codeql.controlflow.BasicBlock as BB
private import codeql.util.Unit
private import codeql.controlflow.SuccessorType

Instruction getASuccessor(Instruction i) { result = i.getASuccessor() }

Instruction getAPredecessor(Instruction i) { i = getASuccessor(result) }

private predicate isJoin(Instruction i) { strictcount(getAPredecessor(i)) > 1 }

private predicate isBranch(Instruction i) { strictcount(getASuccessor(i)) > 1 }

private predicate startsBasicBlock(Instruction i) {
  i instanceof ProgramEntryInstruction
  or
  not exists(getAPredecessor(i)) and exists(getASuccessor(i))
  or
  any(Call call).getTarget() = i
  or
  isJoin(i)
  or
  isBranch(getAPredecessor(i))
}

newtype TBasicBlock = TMkBasicBlock(Instruction i) { startsBasicBlock(i) }

private predicate intraBBSucc(Instruction i1, Instruction i2) {
  i2 = getASuccessor(i1) and
  not startsBasicBlock(i2)
}

private predicate bbIndex(Instruction bbStart, Instruction i, int index) =
  shortestDistances(startsBasicBlock/1, intraBBSucc/2)(bbStart, i, index)

private predicate entryBB(FunctionEntryBasicBlock entry) { any() }

private predicate succBB(BasicBlock pred, BasicBlock succ) { pred.getASuccessor() = succ }

/** Holds if `dom` is an immediate dominator of `bb`. */
cached
private predicate bbIdominates(BasicBlock dom, BasicBlock bb) =
  idominance(entryBB/1, succBB/2)(_, dom, bb)

private predicate predBB(BasicBlock succ, BasicBlock pred) { pred.getASuccessor() = succ }

class FunctionExitBasicBlock extends BasicBlock {
  FunctionExitBasicBlock() { this.getLastInstruction() instanceof Ret }
}

private predicate exitBB(FunctionExitBasicBlock exit) { any() }

/** Holds if `dom` is an immediate post-dominator of `bb`. */
cached
predicate bbIPostDominates(BasicBlock dom, BasicBlock bb) =
  idominance(exitBB/1, predBB/2)(_, dom, bb)

class BasicBlock extends TBasicBlock {
  Instruction getInstruction(int index) { bbIndex(this.getFirstInstruction(), result, index) }

  Instruction getAnInstruction() { result = this.getInstruction(_) }

  Instruction getFirstInstruction() { this = TMkBasicBlock(result) }

  Instruction getLastInstruction() {
    result = this.getInstruction(this.getNumberOfInstructions() - 1)
  }

  BasicBlock getASuccessor() {
    result.getFirstInstruction() = this.getLastInstruction().getASuccessor()
  }

  BasicBlock getBackEdgeSuccessor() {
    result = this.getASuccessor() and
    result.getFirstInstruction().getIndex() < this.getFirstInstruction().getIndex()
  }

  BasicBlock getAPredecessor() { this = result.getASuccessor() }

  int getNumberOfInstructions() { result = strictcount(this.getInstruction(_)) }

  string toString() {
    result = this.getFirstInstruction().toString() + ".." + this.getLastInstruction()
  }

  string getDumpString() {
    result =
      strictconcat(int index, Instruction instr |
        instr = this.getInstruction(index)
      |
        instr.toString(), "\n" order by index
      )
  }

  Location getLocation() { result = this.getFirstInstruction().getLocation() }

  Function getEnclosingFunction() { result.getABasicBlock() = this }

  predicate strictlyDominates(BasicBlock bb) { bbIdominates+(this, bb) }

  predicate dominates(BasicBlock bb) { this.strictlyDominates(bb) or this = bb }

  predicate inDominanceFrontier(BasicBlock df) {
    this.getASuccessor() = df and
    not bbIdominates(this, df)
    or
    exists(BasicBlock prev | prev.inDominanceFrontier(df) |
      bbIdominates(this, prev) and
      not bbIdominates(this, df)
    )
  }

  BasicBlock getImmediateDominator() { bbIdominates(result, this) }

  /**
   * Holds if this basic block strictly post-dominates basic block `bb`.
   *
   * That is, all paths reaching a normal exit point basic block from basic
   * block `bb` must go through this basic block and this basic block is
   * different from `bb`.
   */
  predicate strictlyPostDominates(BasicBlock bb) { bbIPostDominates+(this, bb) }

  /**
   * Holds if this basic block post-dominates basic block `bb`.
   *
   * That is, all paths reaching a normal exit point basic block from basic
   * block `bb` must go through this basic block.
   */
  predicate postDominates(BasicBlock bb) {
    this.strictlyPostDominates(bb) or
    this = bb
  }
}

final private class BasicBlockAlias = BasicBlock;

class ProgramEntryBasicBlock extends BasicBlock {
  ProgramEntryBasicBlock() { this.getFirstInstruction() instanceof ProgramEntryInstruction }
}

module BinaryCfg implements BB::CfgSig<Location> {
  class ControlFlowNode = Instruction;

  class BasicBlock extends BasicBlockAlias {
    ControlFlowNode getNode(int i) { result = super.getInstruction(i) }

    ControlFlowNode getLastNode() { result = super.getLastInstruction() }

    int length() { result = strictcount(super.getInstruction(_)) }

    BasicBlock getASuccessor() { result = super.getASuccessor() }

    BasicBlock getASuccessor(SuccessorType t) { exists(t) and result = super.getASuccessor() }

    predicate strictlyDominates(BasicBlock bb) { super.strictlyDominates(bb) }

    predicate dominates(BasicBlock bb) { super.dominates(bb) }

    BasicBlock getImmediateDominator() { result = super.getImmediateDominator() }

    predicate inDominanceFrontier(BasicBlock df) { super.inDominanceFrontier(df) }

    predicate strictlyPostDominates(BasicBlock bb) { super.strictlyPostDominates(bb) }

    predicate postDominates(BasicBlock bb) { super.postDominates(bb) }
  }

  class EntryBasicBlock extends BasicBlock instanceof FunctionEntryBasicBlock { }

  pragma[nomagic]
  predicate dominatingEdge(BasicBlock bb1, BasicBlock bb2) {
    bb1.getASuccessor() = bb2 and
    bb1 = bb2.getImmediateDominator() and
    forall(BasicBlock pred | pred = bb2.getAPredecessor() and pred != bb1 | bb2.dominates(pred))
  }
}
