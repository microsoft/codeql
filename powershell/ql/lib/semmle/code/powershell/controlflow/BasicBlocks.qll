/** Provides classes representing basic blocks. */

private import powershell
private import ControlFlowGraph
private import CfgNodes
private import SuccessorTypes
private import internal.ControlFlowGraphImpl as CfgImpl
private import CfgImpl::BasicBlocks as BasicBlocksImpl

/**
 * A basic block, that is, a maximal straight-line sequence of control flow nodes
 * without branches or joins.
 */
final class BasicBlock extends BasicBlocksImpl::BasicBlock {
  /** Gets an immediate successor of this basic block, if any. */
  BasicBlock getASuccessor() { result = super.getASuccessor() }

  /** Gets an immediate successor of this basic block of a given type, if any. */
  BasicBlock getASuccessor(SuccessorType t) { result = super.getASuccessor(t) }

  /** Gets an immediate predecessor of this basic block, if any. */
  BasicBlock getAPredecessor() { result = super.getAPredecessor() }

  /** Gets an immediate predecessor of this basic block of a given type, if any. */
  BasicBlock getAPredecessor(SuccessorType t) { result = super.getAPredecessor(t) }

  // The overrides below are to use `CfgNode` instead of `CfgImpl::Node`
  CfgNode getNode(int pos) { result = super.getNode(pos) }

  CfgNode getANode() { result = super.getANode() }

  /** Gets the first control flow node in this basic block. */
  CfgNode getFirstNode() { result = super.getFirstNode() }

  /** Gets the last control flow node in this basic block. */
  CfgNode getLastNode() { result = super.getLastNode() }

  /**
   * Holds if this basic block immediately dominates basic block `bb`.
   *
   * That is, this basic block is the unique basic block satisfying:
   * 1. This basic block strictly dominates `bb`
   * 2. There exists no other basic block that is strictly dominated by this
   *    basic block and which strictly dominates `bb`.
   *
   * All basic blocks, except entry basic blocks, have a unique immediate
   * dominator.
   */
  predicate immediatelyDominates(BasicBlock bb) { super.immediatelyDominates(bb) }

  /**
   * Holds if this basic block strictly dominates basic block `bb`.
   *
   * That is, all paths reaching basic block `bb` from some entry point
   * basic block must go through this basic block (which must be different
   * from `bb`).
   */
  predicate strictlyDominates(BasicBlock bb) { super.strictlyDominates(bb) }

  /**
   * Holds if this basic block dominates basic block `bb`.
   *
   * That is, all paths reaching basic block `bb` from some entry point
   * basic block must go through this basic block.
   */
  predicate dominates(BasicBlock bb) { super.dominates(bb) }

  /**
   * Holds if `df` is in the dominance frontier of this basic block.
   * That is, this basic block dominates a predecessor of `df`, but
   * does not dominate `df` itself.
   */
  predicate inDominanceFrontier(BasicBlock df) { super.inDominanceFrontier(df) }

  /**
   * Gets the basic block that immediately dominates this basic block, if any.
   *
   * That is, the result is the unique basic block satisfying:
   * 1. The result strictly dominates this basic block.
   * 2. There exists no other basic block that is strictly dominated by the
   *    result and which strictly dominates this basic block.
   *
   * All basic blocks, except entry basic blocks, have a unique immediate
   * dominator.
   */
  BasicBlock getImmediateDominator() { result = super.getImmediateDominator() }

  /**
   * Holds if the edge with successor type `s` out of this basic block is a
   * dominating edge for `dominated`.
   *
   * That is, all paths reaching `dominated` from the entry point basic
   * block must go through the `s` edge out of this basic block.
   *
   * Edge dominance is similar to node dominance except it concerns edges
   * instead of nodes: A basic block is dominated by a _basic block_ `bb` if it
   * can only be reached through `bb` and dominated by an _edge_ `s` if it can
   * only be reached through `s`.
   *
   * Note that where all basic blocks (except the entry basic block) are
   * strictly dominated by at least one basic block, a basic block may not be
   * dominated by any edge. If an edge dominates a basic block `bb`, then
   * both endpoints of the edge dominates `bb`. The converse is not the case,
   * as there may be multiple paths between the endpoints with none of them
   * dominating.
   */
  predicate edgeDominates(BasicBlock dominated, SuccessorType s) {
    super.edgeDominates(dominated, s)
  }

  /**
   * Holds if this basic block strictly post-dominates basic block `bb`.
   *
   * That is, all paths reaching a normal exit point basic block from basic
   * block `bb` must go through this basic block (which must be different
   * from `bb`).
   */
  predicate strictlyPostDominates(BasicBlock bb) { super.strictlyPostDominates(bb) }

  /**
   * Holds if this basic block post-dominates basic block `bb`.
   *
   * That is, all paths reaching a normal exit point basic block from basic
   * block `bb` must go through this basic block.
   */
  predicate postDominates(BasicBlock bb) { super.postDominates(bb) }
}

/**
 * An entry basic block, that is, a basic block whose first node is
 * an entry node.
 */
final class EntryBasicBlock extends BasicBlock, BasicBlocksImpl::EntryBasicBlock { }

/**
 * An annotated exit basic block, that is, a basic block that contains an
 * annotated exit node.
 */
final class AnnotatedExitBasicBlock extends BasicBlock, BasicBlocksImpl::AnnotatedExitBasicBlock { }

/**
 * An exit basic block, that is, a basic block whose last node is
 * an exit node.
 */
final class ExitBasicBlock extends BasicBlock, BasicBlocksImpl::ExitBasicBlock { }

/** A basic block with more than one predecessor. */
final class JoinBlock extends BasicBlock, BasicBlocksImpl::JoinBasicBlock {
  JoinBlockPredecessor getJoinBlockPredecessor(int i) { result = super.getJoinBlockPredecessor(i) }
}

/** A basic block that is an immediate predecessor of a join block. */
final class JoinBlockPredecessor extends BasicBlock, BasicBlocksImpl::JoinPredecessorBasicBlock { }

/**
 * A basic block that terminates in a condition, splitting the subsequent
 * control flow.
 */
final class ConditionBlock extends BasicBlock, BasicBlocksImpl::ConditionBasicBlock { }
