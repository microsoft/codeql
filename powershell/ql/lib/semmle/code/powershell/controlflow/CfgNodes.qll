/** Provides classes representing nodes in a control flow graph. */

private import powershell
private import BasicBlocks
private import ControlFlowGraph
private import internal.ControlFlowGraphImpl as CfgImpl

/** An entry node for a given scope. */
class EntryNode extends CfgNode, CfgImpl::EntryNode {
  override string getAPrimaryQlClass() { result = "EntryNode" }

  final override EntryBasicBlock getBasicBlock() { result = super.getBasicBlock() }
}

/** An exit node for a given scope, annotated with the type of exit. */
class AnnotatedExitNode extends CfgNode, CfgImpl::AnnotatedExitNode {
  override string getAPrimaryQlClass() { result = "AnnotatedExitNode" }

  final override AnnotatedExitBasicBlock getBasicBlock() { result = super.getBasicBlock() }
}

/** An exit node for a given scope. */
class ExitNode extends CfgNode, CfgImpl::ExitNode {
  override string getAPrimaryQlClass() { result = "ExitNode" }
}

/**
 * A node for an AST node.
 *
 * Each AST node maps to zero or more `AstCfgNode`s: zero when the node is unreachable
 * (dead) code or not important for control flow, and multiple when there are different
 * splits for the AST node.
 */
class AstCfgNode extends CfgNode, CfgImpl::AstCfgNode {
  /** Gets the name of the primary QL class for this node. */
  override string getAPrimaryQlClass() { result = "AstCfgNode" }
}

/** A control-flow node that wraps an AST expression. */
class ExprCfgNode extends AstCfgNode {
  override string getAPrimaryQlClass() { result = "ExprCfgNode" }

  Expr e;

  ExprCfgNode() { e = this.getAstNode() }

  /** Gets the underlying expression. */
  Expr getExpr() { result = e }
}

/** A control-flow node that wraps an AST statement. */
class StmtCfgNode extends AstCfgNode {
  override string getAPrimaryQlClass() { result = "StmtCfgNode" }

  Stmt s;

  StmtCfgNode() { s = this.getAstNode() }

  /** Gets the underlying expression. */
  Stmt getStmt() { result = s }
}

/**
 * A class for mapping parent-child AST nodes to parent-child CFG nodes.
 */
abstract private class ChildMapping extends Ast {
  /**
   * Holds if `child` is a (possibly nested) child of this expression
   * for which we would like to find a matching CFG child.
   */
  abstract predicate relevantChild(Ast child);

  pragma[nomagic]
  abstract predicate reachesBasicBlock(Ast child, CfgNode cfn, BasicBlock bb);

  /**
   * Holds if there is a control-flow path from `cfn` to `cfnChild`, where `cfn`
   * is a control-flow node for this expression, and `cfnChild` is a control-flow
   * node for `child`.
   *
   * The path never escapes the syntactic scope of this expression.
   */
  cached
  predicate hasCfgChild(Ast child, CfgNode cfn, CfgNode cfnChild) {
    this.reachesBasicBlock(child, cfn, cfnChild.getBasicBlock()) and
    cfnChild.getAstNode() = child
  }
}

/**
 * A class for mapping parent-child AST nodes to parent-child CFG nodes.
 */
abstract private class ExprChildMapping extends Expr, ChildMapping {
  pragma[nomagic]
  override predicate reachesBasicBlock(Ast child, CfgNode cfn, BasicBlock bb) {
    this.relevantChild(child) and
    cfn.getAstNode() = this and
    bb.getANode() = cfn
    or
    exists(BasicBlock mid |
      this.reachesBasicBlock(child, cfn, mid) and
      bb = mid.getAPredecessor() and
      not mid.getANode().getAstNode() = child
    )
  }
}

/**
 * A class for mapping parent-child AST nodes to parent-child CFG nodes.
 */
abstract private class NonExprChildMapping extends ChildMapping {
  NonExprChildMapping() { not this instanceof Expr }

  pragma[nomagic]
  override predicate reachesBasicBlock(Ast child, CfgNode cfn, BasicBlock bb) {
    this.relevantChild(child) and
    cfn.getAstNode() = this and
    bb.getANode() = cfn
    or
    exists(BasicBlock mid |
      this.reachesBasicBlock(child, cfn, mid) and
      bb = mid.getASuccessor() and
      not mid.getANode().getAstNode() = child
    )
  }
}

/** Provides classes for control-flow nodes that wrap AST expressions. */
module ExprNodes {
  private class VarAccessChildMapping extends ExprChildMapping, VarAccess {
    override predicate relevantChild(Ast n) { none() }
  }

  class VarAccessCfgNode extends ExprCfgNode {
    override string getAPrimaryQlClass() { result = "VarAccessCfgNode" }

    override VarAccessChildMapping e;

    override VarAccess getExpr() { result = super.getExpr() }
  }

  private class VarReadAccessChildMapping extends VarAccessChildMapping, VarReadAccess { }

  class VarReadAccessCfgNode extends VarAccessCfgNode {
    override string getAPrimaryQlClass() { result = "VarReadAccessCfgNode" }

    override VarReadAccessChildMapping e;

    override VarReadAccess getExpr() { result = super.getExpr() }
  }

  private class VarWriteAccessChildMapping extends VarAccessChildMapping, VarWriteAccess { }

  class VarWriteAccessCfgNode extends VarAccessCfgNode {
    override string getAPrimaryQlClass() { result = "VarWriteAccessCfgNode" }

    override VarWriteAccessChildMapping e;

    override VarWriteAccess getExpr() { result = super.getExpr() }

    Variable getVariable() { result = e.getVariable() }

    predicate isExplicitWrite(StmtNodes::AssignStmtCfgNode assignment) {
      this = assignment.getLeftHandSide()
    }

    predicate isImplicitWrite() { e.isImplicit() }
  }

  /** A control-flow node that wraps an argument expression. */
  class ArgumentCfgNode extends ExprCfgNode {
    override string getAPrimaryQlClass() { result = "ArgumentCfgNode" }

    override Argument e;

    final override Argument getExpr() { result = super.getExpr() }
  }

  private class InvokeMemberChildMapping extends ExprChildMapping, InvokeMemberExpr {
    override predicate relevantChild(Ast n) { n = this.getQualifier() or n = this.getAnArgument() }
  }

  /** A control-flow node that wraps an `InvokeMemberExpr` expression. */
  class InvokeMemberCfgNode extends ExprCfgNode {
    override string getAPrimaryQlClass() { result = "InvokeMemberCfgNode" }

    override InvokeMemberChildMapping e;

    final override InvokeMemberExpr getExpr() { result = super.getExpr() }

    final ExprCfgNode getQualifier() { e.hasCfgChild(e.getQualifier(), this, result) }
  }

  /** A control-flow node that wraps a qualifier expression. */
  class QualifierCfgNode extends ExprCfgNode {
    QualifierCfgNode() { this = any(InvokeMemberCfgNode invoke).getQualifier() }

    InvokeMemberCfgNode getInvokeMember() { this = result.getQualifier() }
  }

  class ConditionalChildMapping extends ExprChildMapping, ConditionalExpr {
    override predicate relevantChild(Ast n) { n = this.getCondition() or n = this.getABranch() }
  }

  /** A control-flow node that wraps a `ConditionalExpr` expression. */
  class ConditionalCfgNode extends ExprCfgNode {
    override string getAPrimaryQlClass() { result = "ConditionalCfgNode" }

    override ConditionalChildMapping e;

    final override ConditionalExpr getExpr() { result = super.getExpr() }

    final ExprCfgNode getCondition() { e.hasCfgChild(e.getCondition(), this, result) }

    final ExprCfgNode getBranch(boolean value) { e.hasCfgChild(e.getBranch(value), this, result) }

    final ExprCfgNode getABranch() { result = this.getBranch(_) }

    final ExprCfgNode getIfTrue() { e.hasCfgChild(e.getIfTrue(), this, result) }

    final ExprCfgNode getIfFalse() { e.hasCfgChild(e.getIfFalse(), this, result) }
  }

  class MemberChildMapping extends ExprChildMapping, MemberExpr {
    override predicate relevantChild(Ast n) { n = this.getBase() or n = this.getMember() }
  }

  /** A control-flow node that wraps a `MemberExpr` expression. */
  class MemberCfgNode extends ExprCfgNode {
    override string getAPrimaryQlClass() { result = "MemberCfgNode" }

    override MemberChildMapping e;

    final override MemberExpr getExpr() { result = super.getExpr() }

    final ExprCfgNode getBase() { e.hasCfgChild(e.getBase(), this, result) }

    final string getMemberName() { result = e.getMemberName() }

    predicate isStatic() { e.isStatic() }
  }

  /** A control-flow node that wraps a `MemberExpr` expression that is being written to. */
  class MemberCfgWriteAccessNode extends MemberCfgNode {
    MemberCfgWriteAccessNode() { this.getExpr() instanceof MemberExprWriteAccess }
  }

  /** A control-flow node that wraps a `MemberExpr` expression that is being read from. */
  class MemberCfgReadAccessNode extends MemberCfgNode {
    MemberCfgReadAccessNode() { this.getExpr() instanceof MemberExprReadAccess }
  }
}

module StmtNodes {
  private class CmdChildMapping extends NonExprChildMapping, Cmd {
    override predicate relevantChild(Ast n) { n = this.getElement(_) }
  }

  /** A control-flow node that wraps a `Cmd` AST expression. */
  class CmdCfgNode extends StmtCfgNode {
    override string getAPrimaryQlClass() { result = "CmdCfgNode" }

    override CmdChildMapping s;

    override Cmd getStmt() { result = super.getStmt() }
  }

  private class AssignStmtChildMapping extends NonExprChildMapping, AssignStmt {
    override predicate relevantChild(Ast n) {
      n = this.getLeftHandSide() or n = this.getRightHandSide()
    }
  }

  /** A control-flow node that wraps an `AssignStmt` AST expression. */
  class AssignStmtCfgNode extends StmtCfgNode {
    override string getAPrimaryQlClass() { result = "AssignCfgNode" }

    override AssignStmtChildMapping s;

    override AssignStmt getStmt() { result = super.getStmt() }

    /** Gets the LHS of this assignment. */
    final ExprCfgNode getLeftHandSide() { s.hasCfgChild(s.getLeftHandSide(), this, result) }

    /** Gets the RHS of this assignment. */
    final StmtCfgNode getRightHandSide() { s.hasCfgChild(s.getRightHandSide(), this, result) }
  }
}
