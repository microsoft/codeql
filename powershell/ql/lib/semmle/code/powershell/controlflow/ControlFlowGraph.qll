/** Provides classes representing the control flow graph. */

private import powershell
private import BasicBlocks
private import SuccessorTypes
private import internal.ControlFlowGraphImpl as CfgImpl
private import internal.Splitting as Splitting
private import internal.Completion

/**
 * An AST node with an associated control-flow graph.
 *
 * Top-levels, methods, blocks, and lambdas are all CFG scopes.
 *
 * Note that module declarations are not themselves CFG scopes, as they are part of
 * the CFG of the enclosing top-level or callable.
 */
class CfgScope extends Scope instanceof CfgImpl::CfgScope {
  final CfgScope getOuterCfgScope() {
    exists(Ast parent |
      parent = this.getParent() and
      result = CfgImpl::getCfgScope(parent)
    )
  }

  Parameter getAParameter() { result = super.getAParameter() }
}

/**
 * A control flow node.
 *
 * A control flow node is a node in the control flow graph (CFG). There is a
 * many-to-one relationship between CFG nodes and AST nodes.
 *
 * Only nodes that can be reached from an entry point are included in the CFG.
 */
class CfgNode extends CfgImpl::Node {
  /** Gets the name of the primary QL class for this node. */
  string getAPrimaryQlClass() { none() }

  /** Gets the file of this control flow node. */
  final File getFile() { result = this.getLocation().getFile() }

  /** Gets a successor node of a given type, if any. */
  final CfgNode getASuccessor(SuccessorType t) { result = super.getASuccessor(t) }

  /** Gets an immediate successor, if any. */
  final CfgNode getASuccessor() { result = this.getASuccessor(_) }

  /** Gets an immediate predecessor node of a given flow type, if any. */
  final CfgNode getAPredecessor(SuccessorType t) { result.getASuccessor(t) = this }

  /** Gets an immediate predecessor, if any. */
  final CfgNode getAPredecessor() { result = this.getAPredecessor(_) }

  /** Gets the basic block that this control flow node belongs to. */
  BasicBlock getBasicBlock() { result.getANode() = this }
}

/** The type of a control flow successor. */
class SuccessorType extends CfgImpl::TSuccessorType {
  /** Gets a textual representation of successor type. */
  string toString() { none() }
}

/** Provides different types of control flow successor types. */
module SuccessorTypes {
  /** A normal control flow successor. */
  class NormalSuccessor extends SuccessorType, CfgImpl::TSuccessorSuccessor {
    final override string toString() { result = "successor" }
  }

  /**
   * A conditional control flow successor. Either a Boolean successor (`BooleanSuccessor`)
   * or a matching successor (`MatchingSuccessor`)
   */
  abstract class ConditionalSuccessor extends SuccessorType {
    boolean value;

    bindingset[value]
    ConditionalSuccessor() { any() }

    /** Gets the Boolean value of this successor. */
    final boolean getValue() { result = value }

    override string toString() { result = this.getValue().toString() }
  }

  class BooleanSuccessor extends ConditionalSuccessor, CfgImpl::TBooleanSuccessor {
    BooleanSuccessor() { this = CfgImpl::TBooleanSuccessor(value) }
  }

  class MatchingSuccessor extends ConditionalSuccessor, CfgImpl::TMatchingSuccessor {
    MatchingSuccessor() { this = CfgImpl::TMatchingSuccessor(value) }
  }

  class ReturnSuccessor extends SuccessorType, CfgImpl::TReturnSuccessor {
    final override string toString() { result = "return" }
  }

  class BreakSuccessor extends SuccessorType, CfgImpl::TBreakSuccessor {
    final override string toString() { result = "break" }
  }

  class ContinueSuccessor extends SuccessorType, CfgImpl::TContinueSuccessor {
    final override string toString() { result = "continue" }
  }

  class ThrowSuccessor extends SuccessorType, CfgImpl::TThrowSuccessor {
    final override string toString() { result = "throw" }
  }

  class ExitSuccessor extends SuccessorType, CfgImpl::TExitSuccessor {
    final override string toString() { result = "exit" }
  }

  class EmptinessSuccessor extends ConditionalSuccessor, CfgImpl::TEmptinessSuccessor {
    EmptinessSuccessor() { this = CfgImpl::TEmptinessSuccessor(value) }

    /** Holds if this is an empty successor. */
    predicate isEmpty() { value = true }

    override string toString() { if this.isEmpty() then result = "empty" else result = "non-empty" }
  }
}

class Split = Splitting::Split;

/** Provides different kinds of control flow graph splittings. */
module Split {
  class ConditionalCompletionSplit = Splitting::ConditionalCompletionSplit;
}
