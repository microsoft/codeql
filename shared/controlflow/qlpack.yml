name: codeql/controlflow
version: 2.0.10
groups: shared
library: true
dependencies:
  codeql/util: ${workspace}
warnOnImplicitThis: true
