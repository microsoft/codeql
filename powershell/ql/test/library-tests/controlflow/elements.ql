import semmle.code.powershell.controlflow.CfgNodes
import ExprNodes
import StmtNodes

query predicate forEach(ForEachStmtCfgNode forEach, ExprCfgNode va, ExprCfgNode iterable, StmtCfgNode body) {
  va = forEach.getVarAccess() and
  iterable = forEach.getIterableExpr() and
  body = forEach.getBody()
}