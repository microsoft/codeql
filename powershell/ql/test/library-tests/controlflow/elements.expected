| graph/functions.ps1:29:5:32:5 | forach(... in ...) | graph/functions.ps1:29:14:29:20 | number | graph/functions.ps1:29:25:29:32 | numbers | graph/functions.ps1:29:35:32:5 | {...} |
| graph/loops.ps1:52:5:55:5 | forach(... in ...) | graph/loops.ps1:52:14:52:20 | letter | graph/loops.ps1:52:25:52:36 | letterArray | graph/loops.ps1:53:5:55:5 | {...} |
